spec/ics-010-grandpa-client/README.md

```diff
@@ -0,0 +1,345 @@
+---
+ics: 10
+title: GRANDPA Client
+stage: draft
+category: IBC/TAO
+kind: instantiation
+author: Yuanchao Sun <ys@cdot.network>, John Wu <john@cdot.network>
+created: 2020-03-15
+implements: 2
+---
+
+## Synopsis
+
+This specification document describes a client (verification algorithm) for a blockchain using GRANDPA.
+
+GRANDPA (GHOST-based Recursive ANcestor Deriving Prefix Agreement) is a finality gadget that will be used by the Polkadot relay chain. It now has a Rust implementation and is part of the Substrate, so likely blockchains built using Substrate will use GRANDPA as its finality gadget.
+
+### Motivation
+
+Blockchains using GRANDPA finality gadget might like to interface with other replicated state machines or solo machines over IBC.
+
+### Definitions
+
+Functions & terms are as defined in [ICS 2](../ics-002-client-semantics).
+
+### Desired Properties
+
+This specification must satisfy the client interface defined in ICS 2.
+
+## Technical Specification
+
+This specification depends on correct instantiation of the [GRANDPA finality gadget](https://github.com/w3f/consensus/blob/master/pdf/grandpa.pdf) and its light client algorithm.
+
+### Client state
+
+The GRANDPA client state tracks latest height and a possible frozen height.
+
+```typescript
+interface ClientState {
+  latestHeight: uint64
+  frozenHeight: Maybe<uint64>
+}
+```
+
+### Authority set
+
+A set of authorities for GRANDPA.
+
+```typescript
+interface AuthoritySet {
+  // this is incremented every time the set changes
+  setId: uint64
+  authorities: List<Pair<AuthorityId, AuthorityWeight>>
+}
+```
+
+### Consensus state
+
+The GRANDPA client tracks authority set and commitment root for all previously verified consensus states.
+
+```typescript
+interface ConsensusState {
+  authoritySet: AuthoritySet
+  commitmentRoot: []byte
+}
+```
+
+### Headers
+
+The GRANDPA client headers include the height, the commitment root, a justification of block and a proof of authority set.
+
+```typescript
+interface Header {
+  height: uint64
+  commitmentRoot: []byte
+  justification: Justification
+  authoritySetProof: []byte
+}
+```
+
+### Justification
+
+A GRANDPA justification for block finality, it includes a commit message and an ancestry proof including all headers routing all precommit target blocks to the commit target block.
+
+```typescript
+interface Justification {
+  round: uint64
+  commit: Commit
+  votesAncestries: []Header
+}
+```
+
+### Evidence
+
+The `Evidence` type is used for detecting misbehaviour and freezing the client - to prevent further packet flow - if applicable.
+GRANDPA client `Evidence` consists of two headers at the same height both of which the light client would have considered valid.
```

Comment:
Given a particular previous header which was already trusted?

Reply:
I don't think a previous trusted header is needed, the current consensus state should be enough to verify whether these two headers are valid.

Reply:
Right, my question is whether you also want to support equivocation from a previous header? See the discussion in ICS 7 about "would-have-been-fooled".